Allow running from any directory in the repo

gkotian opened the pull request:
Deal only with absolute file paths from the get-go. If considering all
modified files, then construct the file paths using `GIT_TOP`; but if
considering only modified files given by the user, then construct the
file paths using `readlink -m`.

Fixes #15.

gkotian committed:
Allow running from any directory in the repo

Deal only with absolute file paths from the get-go. If considering all
modified files, then construct the file paths using `GIT_TOP`; but if
considering only modified files given by the user, then construct the
file paths using `readlink -m`.

Fixes #15.

gkotian commented:
This is an alternative to #16.

gkotian commented:
@crash5 could you please review this, and check if it covers all your cases?

crash5 commented:
Hi!

If call script simple with `git_add_gui.sh` it works well but call with git alias can't handle the sub directories.

Reproduce:

Set alias like `git config --global alias.patch-test '!git_add_gui.sh'`


Setup test repository:
```bash
git init
mkdir -p 1st/2nd/3rd
echo "root" >> readme.txt
echo -n 1st{,/2nd{,/3rd}} | xargs -d' ' -i sh -c 'echo "$1" > "$1"/readme.txt' -- {}
git add .
git commit -m "Initial"
echo -n 1st{,/2nd{,/3rd}} | xargs -d' ' -i sh -c 'echo "$1" >> "$1"/readme.txt' -- {}
```

Produce this structure
```
|-.git
|-1st
|  |-2nd
|  |  |-3rd
|  |  |  |-readme.txt
|  |  |-readme.txt
|  |-readme.txt
|-readme.txt
```

With modified readme files in the subdirectories (readme.txt in root is not modified)
```git
## master
 M 1st/2nd/3rd/readme.txt
 M 1st/2nd/readme.txt
 M 1st/readme.txt
```

Run `git patch-test readme.txt` from `1st/` directory. Script open up diff editor with `readme.txt` from the `root` directory instead of `1st`.

gkotian commented:
I tried running the command from various directories in the hierarchy, and the `FILES_LIST` array always seems to be correct:
```
## (master) [~/tmp_home/pull_18] > pwd
/home/gautam/tmp_home/pull_18
## (master) [~/tmp_home/pull_18] > git status -s
 M 1st/2nd/3rd/readme.txt
 M 1st/2nd/readme.txt
 M 1st/readme.txt                                                   
## (master) [~/tmp_home/pull_18] > git patch-test                   
GIT_TOP = /home/gautam/tmp_home/pull_18                             
GIT_PREFIX =                                                        
FILES_LIST:                                                         
    /home/gautam/tmp_home/pull_18/1st/2nd/3rd/readme.txt            
    /home/gautam/tmp_home/pull_18/1st/2nd/readme.txt                
    /home/gautam/tmp_home/pull_18/1st/readme.txt                    
                                                                    
                                                                    
## (master) [~/tmp_home/pull_18] > cd 1st                           
## (master) [~/tmp_home/pull_18/1st] > git patch-test               
GIT_TOP = /home/gautam/tmp_home/pull_18                             
GIT_PREFIX = 1st/                                                   
FILES_LIST:                                                         
    /home/gautam/tmp_home/pull_18/1st/2nd/3rd/readme.txt            
    /home/gautam/tmp_home/pull_18/1st/2nd/readme.txt                
    /home/gautam/tmp_home/pull_18/1st/readme.txt                    
                                                                    
                                                                    
## (master) [~/tmp_home/pull_18/1st] > cd 2nd                       
## (master) [~/tmp_home/pull_18/1st/2nd] > git patch-test           
GIT_TOP = /home/gautam/tmp_home/pull_18                             
GIT_PREFIX = 1st/2nd/                                               
FILES_LIST:                                                         
    /home/gautam/tmp_home/pull_18/1st/2nd/3rd/readme.txt            
    /home/gautam/tmp_home/pull_18/1st/2nd/readme.txt                
    /home/gautam/tmp_home/pull_18/1st/readme.txt

## (master) [~/tmp_home/pull_18/1st/2nd] > cd 3rd                   
## (master) [~/tmp_home/pull_18/1st/2nd/3rd] > git patch-test       
GIT_TOP = /home/gautam/tmp_home/pull_18                             
GIT_PREFIX = 1st/2nd/3rd/
FILES_LIST:
    /home/gautam/tmp_home/pull_18/1st/2nd/3rd/readme.txt
    /home/gautam/tmp_home/pull_18/1st/2nd/readme.txt
    /home/gautam/tmp_home/pull_18/1st/readme.txt
```

gkotian commented:
Ah shit, I'm an idiot. 

You asked me to run `git patch-test readme.txt` from `1st/` directory, but I've been running only `git patch-test`.

I am able to reproduce the issue now. I'll see if I can come up with a fix.

gkotian committed:
Properly handle passing a list of files manually

When a list of files is manually passed, and the script is launched as a
git alias shell-command, we need to be careful and prefix `GIT_PREFIX`
to the manual file names.

gkotian committed:
Confirm validity of all user-given files

gkotian commented:
Added two more commits that hopefully handle the situation better. @crash5, can you please check if it works for you?

crash5 commented:
Hi!
It looks good. Thx.
